stream-compositions/services/legal-entity-composition-service/src/main/java/com/backbase/stream/compositions/legalentity/core/mapper/GrandCentralPartyMapper.java

```diff
@@ -0,0 +1,124 @@
+package com.backbase.stream.compositions.legalentity.core.mapper;
+
+import com.backbase.grandcentral.event.spec.v1.Data;
+import com.backbase.grandcentral.event.spec.v1.Email;
+import com.backbase.grandcentral.event.spec.v1.PhoneAddress;
+import com.backbase.stream.compositions.legalentity.core.config.GrandCentralPartyDefaultProperties;
+import com.backbase.stream.legalentity.model.EmailAddress;
+import com.backbase.stream.legalentity.model.JobProfileUser;
+import com.backbase.stream.legalentity.model.LegalEntity;
+import com.backbase.stream.legalentity.model.PhoneNumber;
+import com.backbase.stream.legalentity.model.User;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.mapstruct.AfterMapping;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+
+@Mapper(componentModel = "spring")
+@EnableConfigurationProperties(GrandCentralPartyDefaultProperties.class)
+public abstract class GrandCentralPartyMapper {
+
+    @Autowired
+    private GrandCentralPartyDefaultProperties properties;
+
+    @Mapping(source = "partyId", target = "externalId")
+    @Mapping(constant = "CUSTOMER", target = "legalEntityType")
+    public abstract LegalEntity mapLegalEntity(Data party);
+
+    @Mapping(source = "partyId", target = "externalId")
+    public abstract User mapUser(Data party);
+
+    public abstract EmailAddress map(Email email);
+
+    public abstract PhoneNumber map(PhoneAddress phoneAddress);
+
+    @AfterMapping
+    protected void updateLegalEntity(Data party, @MappingTarget LegalEntity legalEntity) {
+        if (properties.getParentExternalId() == null
+            || properties.getReferenceJobRoleNames() == null
+            || properties.getReferenceJobRoleNames().isEmpty()) {
+            throw new RuntimeException("Parent external id or reference job role names is empty");
+        }
+
+        legalEntity.setName(getPartyName(party));
+        legalEntity.setRealmName(properties.getRealmName());
+        legalEntity.setParentExternalId(properties.getParentExternalId());
+        if (Data.PartyType.PERSON.equals(party.getPartyType())) {
+            legalEntity.setUsers(List.of(new JobProfileUser(mapUser(party))
+                .referenceJobRoleNames(properties.getReferenceJobRoleNames())));
+        } else if (party.getCustomFields() != null) {
+            // If a party is not a person, the custom fields should be applicable for the LE, not the User
+            legalEntity.setAdditions(party.getCustomFields().getAdditions());
```

Comment:
This is not how the custom field works. The SSDK maven plugin set the `additionalProperties: false` for every event it generates and adds the `additions` property to all objects. There is a ticket to fix this.

Reply:
in the meantime, I have implemented a ~~nasty~~ workaround to generate this guy properly:
```json
     "customFields": {
        "type": "object",
        "description": "Hosts custom fields related to party",
        "existingJavaType": "java.util.Map<String,Object>"
      },
```